odoo/models.py

```diff
@@ -3731,18 +3741,76 @@ def _write(self, vals):
                     if pleft0 < pleft1 <= pright0:
                         raise UserError(_('Recursivity Detected.'))
 
-                    # make some room for parent_left and parent_right at the new position
-                    cr.execute('UPDATE %s SET parent_left=parent_left+%%s WHERE %%s<=parent_left' % self._table, (width, pleft1))
-                    cr.execute('UPDATE %s SET parent_right=parent_right+%%s WHERE %%s<=parent_right' % self._table, (width, pleft1))
-                    # slide the subtree of the current record to its new position
+                    # retrieve the maximum value for parent_right
+                    cr.execute('SELECT MAX(parent_right) FROM %s' % (self._table,))
+                    max_parent_right = cr.fetchone()[0] or 0
+
+                    # move the subtree to its final position and shift the nodes
+                    # between the initial and final positions; the following
+                    # tricky shifts are necessary to avoid violating the
+                    # uniqueness constraints on parent_left and parent_right
                     if pleft0 < pleft1:
-                        cr.execute('''UPDATE %s SET parent_left=parent_left+%%s, parent_right=parent_right+%%s
-                                      WHERE %%s<=parent_left AND parent_left<%%s''' % self._table,
-                                   (pleft1 - pleft0, pleft1 - pleft0, pleft0, pright0))
+                        pleft1 -= width
+                        # subtree nodes: pleft0 -> pleft0 + 2M -> pleft1
+                        # intermediate nodes: pos -> pos + M -> pos - width
+                        query = """ UPDATE {t}
```

Comment:
No you cannot. I had to fix this because it crashed: shifting `parent_left` has a high chance to introduce duplicates with non-shifted-yet `parent_right` values.

Consider the following example (`parent_left, parent_right` are given in parentheses):
- a node A (41, 44) with a single child B (42, 43).

Imagine that you have to shift that subtree 2 positions to the right in order to make some room for a single node. First, you update `parent_left`, and your nodes become:
- node A (43, 44) and node B (44, 45)

And BOOM, the unique index already rejects that because 44 appears twice!